lib/fast-path.ts

```diff
@@ -347,205 +344,208 @@ FPp.needsParens = function (assumeExpressionContext) {
     return false;
   }
 
-  if (
-    parent.type === "ParenthesizedExpression" ||
-    (node.extra && node.extra.parenthesized)
-  ) {
+  if (parent && parent.type === "ParenthesizedExpression") {
     return false;
   }
 
-  switch (node.type) {
-    case "UnaryExpression":
-    case "SpreadElement":
-    case "SpreadProperty":
-      return (
-        parent.type === "MemberExpression" &&
-        name === "object" &&
-        parent.object === node
-      );
-
-    case "BinaryExpression":
-    case "LogicalExpression":
-      switch (parent.type) {
-        case "CallExpression":
-          return name === "callee" && parent.callee === node;
-
-        case "UnaryExpression":
-        case "SpreadElement":
-        case "SpreadProperty":
-          return true;
+  if (node.extra && node.extra.parenthesized) {
+    return true;
+  }
 
-        case "MemberExpression":
-          return name === "object" && parent.object === node;
+  if (parent) {
```

Comment:
Can you make this say `if (parent) { return false; }` instead?

If I'm reading this right, that would mean the remaining 200 or so lines of this function don't have to get reindented. That would make this diff a lot smaller, and easier to read and review.

It would also make the resulting code easier to read. Instead of being in suspense about what happens if `parent` is falsy -- and having to find the end of this `if` block 194 lines later to find out -- the reader would be able to see the answer immediately. Then they can move on to focusing on the interesting case, without having that question in the back of their head.

Reply:
Ah, yeah, I can. I originally did I think because I thought there was some logic at the bottom that belonged outside the `if(parent)`, but actually it turned out there wasn't.

Reply:
Great :-) And I see it's now been merged!